app/assets/javascripts/discourse/app/lib/hashtag-autocomplete.js

```diff
@@ -154,8 +218,30 @@ function _searchRequest(term, context, resultFunc) {
   return currentSearch;
 }
 
-function _sortedContextParams(context) {
-  return Object.entries(contextBasedParams[context])
-    .sort((a, b) => b[1] - a[1])
-    .map((item) => item[0]);
+function _findAndReplaceSeenHashtagPlaceholder(
+  slug,
+  contextualHashtagConfiguration,
+  hashtagSpan
+) {
+  contextualHashtagConfiguration.forEach((type) => {
+    // remove type suffixes
+    const typePostfix = `::${type}`;
+    if (slug.endsWith(typePostfix)) {
+      slug = slug.slice(0, slug.length - typePostfix.length);
+    }
+
+    // replace raw span for the hashtag with a cooked one
+    const matchingSeenHashtag = seenHashtags[type]?.[slug];
+    if (matchingSeenHashtag) {
+      // NOTE: When changing the HTML structure here, you must also change
+      // it in the hashtag-autocomplete markdown rule, and vice-versa.
+      const link = document.createElement("a");
+      link.classList.add("hashtag-cooked");
+      link.href = matchingSeenHashtag.relative_url;
+      link.dataset.type = type;
+      link.dataset.slug = matchingSeenHashtag.slug;
+      link.innerHTML = `<span><svg class="fa d-icon d-icon-${matchingSeenHashtag.icon} svg-icon svg-node"><use href="#${matchingSeenHashtag.icon}"></use></svg>${matchingSeenHashtag.text}</span>`;
```

Comment:
Should we consider escaping any of these values?

Reply:
Hmm I don't think so, we didn't in the old way here https://github.com/discourse/discourse/blob/56d275549b28e1198e3e022bd30220322ab39410/app/assets/javascripts/discourse/app/lib/category-hashtags.js#L6-L11